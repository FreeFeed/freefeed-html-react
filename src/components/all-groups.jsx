/* global CONFIG */
import { memo, useEffect, useMemo, useState, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { sortBy, range } from 'lodash';
import { Link } from 'react-router';
import { Helmet } from 'react-helmet';

import { faCaretDown, faUserFriends } from '@fortawesome/free-solid-svg-icons';
import { getAllGroups } from '../redux/action-creators';
import { Icon } from './fontawesome-icons';

import UserName from './user-name';
import TimeDisplay from './time-display';

import styles from './all-groups.module.scss';
import { UserPicture } from './user-picture';
import { HorScrollable } from './hor-scrollable';

export default function AllGroups() {
  const dispatch = useDispatch();
  const status = useSelector((state) => state.allGroupsStatus);

  useEffect(() => void (status.success || status.loading || dispatch(getAllGroups())), [
    dispatch,
    status,
  ]);

  return (
    <div className="content">
      <Helmet title={`All Groups - ${CONFIG.siteTitle}`} defer={false} />
      <div className="box">
        <div className="box-header-timeline" role="heading">
          All Groups
        </div>
        <div className="box-body">
          <div className="row" style={{ marginBottom: 20 }}>
            <div className="col-md-6">
              <Link to="/groups">List of your groups</Link>
            </div>
            <div className="col-md-6 text-right">
              <Link to="/groups/create">Create a group</Link>
            </div>
          </div>

          {status.loading && <p>Loading...</p>}
          {status.error && (
            <p className="alert alert-danger">Can not load groups list: {status.errorText}</p>
          )}
          {status.success && <GroupsList pageSize={50} />}
        </div>
      </div>
    </div>
  );
}

const SORT_BY_SUBSCRIBERS = 'subscribers';
const SORT_BY_POSTS = 'posts';
const SORT_BY_VARIETY = 'variety';
const SORT_BY_DATE = 'date';
const SORT_DEFAULT = SORT_BY_SUBSCRIBERS;

const sortFields = {
  [SORT_BY_POSTS]: 'postsByMonth',
  [SORT_BY_SUBSCRIBERS]: 'subscribers',
  [SORT_BY_VARIETY]: 'authorsVariety',
  [SORT_BY_DATE]: 'createdAt',
};

function GroupsList({ pageSize }) {
  const [nameFilter, setNameFilter] = useState('');

  const location = useSelector((state) => state.routing.locationBeforeTransitions);
  const { groups, withProtected } = useSelector((state) => state.allGroups);

  const nameFilterLowercase = nameFilter.toLowerCase();
  const filteredGroups = nameFilter
    ? groups.filter((g) => {
        return (
          g.username.toLowerCase().includes(nameFilterLowercase) ||
          g.screenName.toLowerCase().includes(nameFilterLowercase)
        );
      })
    : groups;

  const page = getPageNumber(location);
  const sort = getSortMode(location);
  const totalPages = Math.ceil(filteredGroups.length / pageSize);

  const groupsOnPage = useMemo(
    () =>
      sortBy(filteredGroups, sortFields[sort])
        .reverse()
        .slice(page * pageSize, (page + 1) * pageSize),
    [filteredGroups, page, pageSize, sort],
  );

  return (
    <>
      {withProtected ? (
        <p>
          There are {groups.length} public and protected groups in {CONFIG.siteTitle}. Protected
          groups can only be seen by authenticated users.
        </p>
      ) : (
        <p>
          There are {groups.length} public groups in {CONFIG.siteTitle}.
        </p>
      )}
      <p>
        <label htmlFor="groups-name-filter">Filter by username or display name</label>
        <input
          id="groups-name-filter"
          type="search"
          placeholder="Type to filter"
          onChange={useCallback((e) => setNameFilter(e.target.value), [])}
          className="form-control narrow-input"
        />
      </p>
      <p>Click on the column headers to change table sorting order.</p>

      <HorScrollable>
        <table className={styles.table}>
          <thead>
            <tr className={styles.headersRow}>
              <th className={styles.mainColumn}>Group</th>
              <SortHeader mode={SORT_BY_SUBSCRIBERS} currentMode={sort}>
                Subscribers
              </SortHeader>
              <SortHeader mode={SORT_BY_POSTS} currentMode={sort}>
                Posts per month
              </SortHeader>
              <SortHeader mode={SORT_BY_VARIETY} currentMode={sort}>
                Authors variety
              </SortHeader>
              <SortHeader mode={SORT_BY_DATE} currentMode={sort}>
                Creation date
              </SortHeader>
            </tr>
          </thead>
          <tbody>
            {groupsOnPage.map((g) => (
              <GroupRow key={g.id} g={g} />
            ))}
          </tbody>
        </table>
      </HorScrollable>
      {groupsOnPage.length === 0 ? (
        <div className={styles.empty}>No groups match current filter</div>
      ) : (
        false
      )}
      {totalPages > 1 && (
        <ul className="pagination">
          {range(1, totalPages).map((n) => (
            <li key={n} className={page === n - 1 ? 'active' : undefined}>
              <Link to={linkToPage(n, location)}>{n}</Link>
            </li>
          ))}
        </ul>
      )}
    </>
  );
}

function SortHeader({ children, mode, currentMode }) {
  const location = useSelector((state) => state.routing.locationBeforeTransitions);
  return (
    <th className={styles.sortableHeader + (mode === currentMode ? '' : ` ${styles.inactive}`)}>
      <Link to={linkToSort(mode, location)}>
        <div>
          {children}
          <Icon icon={faCaretDown} className={styles.caret} />
        </div>
      </Link>
    </th>
  );
}

const GroupRow = memo(function GroupRow({ g }) {
  const u = useSelector((state) => state.users[g.id]);
  return (
    <tr className={styles.groupRow}>
      <td>
        <UserPicture user={u} className={styles.userpic} />
        <div className={styles.groupInfo}>
          <UserName user={u}>{u.username}</UserName>
          {u.username !== u.screenName && <div className="small">{u.screenName}</div>}
          {u.isProtected === '1' && (
            <div className="muted-text small">
              <Icon icon={faUserFriends} title="Protected group" />
            </div>
          )}
        </div>
      </td>
      <td>{g.subscribers}</td>
      <td>{g.postsByMonth === 0 ? '< 1' : Math.ceil(g.postsByMonth)}</td>
      <td>{Math.round(g.authorsVariety * 100)}%</td>
      <td>
        <TimeDisplay timeStamp={+u.createdAt} absolute dateOnly />
      </td>
    </tr>
  );
});

function getPageNumber({ query: { page = '1' } }) {
  page = parseInt(page);
  if (!Number.isFinite(page) || page < 1) {
    return 0;
  }
  return page - 1;
}

function getSortMode({ query: { sort = SORT_DEFAULT } }) {
  return sortFields[sort] ? sort : SORT_DEFAULT;
}

function linkToPage(page, { pathname, query }) {
  return {
    pathname,
    query: { ...query, page: page > 1 ? page : undefined },
  };
}

function linkToSort(sort, { pathname }) {
  return {
    pathname,
    query: { sort: sort !== SORT_DEFAULT ? sort : undefined },
  };
}
