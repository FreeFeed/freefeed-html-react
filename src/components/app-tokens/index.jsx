import React from 'react';

import Tokens from './tokens';
import CreateToken from './create-token';
import CreateLink from './create-link';
import ScopesList from './scopes-list';
import Layout from './layout';


export default function ({ routeParams: { splat } }) {
  if (!splat) {
    return <Layout component={Tokens} />;
  }
  if (splat === 'create') {
    return <Layout component={CreateToken} title="Generate new token" />;
  }
  if (splat === 'create-link') {
    return <Layout component={CreateLink} title="Create magic link" />;
  }
  if (splat === 'scopes') {
    return <Layout component={ScopesList} title="Token access rights" />;
  }

  return <p>Page not found: {splat}</p>;
}

