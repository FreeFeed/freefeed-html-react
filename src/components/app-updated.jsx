/* global CONFIG */
import { useSelector } from 'react-redux';
import { useRegisterSW } from 'virtual:pwa-register/react';

import { useEvent } from 'react-use-event-hook';
import { useState } from 'react';
import { isIos } from '../utils/platform-detection';
import styles from './app-updated.module.scss';
import { ButtonLink } from './button-link';

const { intervalSec } = CONFIG.appVersionCheck;

export function AppUpdated() {
  const versionFileUpdated = useSelector((state) => state.appUpdated.updated);
  const [swRegistered, setSwRegistered] = useState(false);

  const {
    needRefresh: [workerUpdated],
    updateServiceWorker,
  } = useRegisterSW({
    onRegistered(r) {
      // iOS has strange behavior with service workers updates, so we will use
      // the old-fashion way (via the version file) here
      if (r && !isIos) {
        setSwRegistered(true);
        setInterval(() => r.update(), intervalSec * 1000);
      }
    },
  });

  const reloadPage = useEvent(() => {
    if (workerUpdated && !isIos) {
      updateServiceWorker();
      // Sometimes the updateServiceWorker doesn't refresh the page, so reload
      // it manually after some time
      setTimeout(() => window.location.reload(true), 2000);
    } else {
      window.location.reload(true);
    }
  });

  const needRefresh = swRegistered ? workerUpdated : versionFileUpdated;

  return needRefresh ? (
    <div className={styles.bar}>
      <div className={styles.indicator}>
        Thereâ€™s an update for {CONFIG.siteTitle}!{' '}
        <ButtonLink className={styles.refresh} onClick={reloadPage}>
          Refresh the page
        </ButtonLink>{' '}
        when you are ready.
      </div>
    </div>
  ) : null;
}
