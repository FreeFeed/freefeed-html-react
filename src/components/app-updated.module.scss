.bar {
  background-color: #337ab7;
  color: white;
}

.refresh {
  color: inherit !important;
  font-weight: bold;
  text-decoration: underline;

  &:hover {
    color: inherit !important;
  }
}

.indicator {
  margin: auto;
  padding: 0.5em 15px;

  @media (max-width: 768px) {
    max-width: 100%;
  }

  @media (min-width: 769px) and (max-width: 992px) {
    max-width: 750px;
  }
  @media (min-width: 993px) {
    max-width: 948px;
  }
}
