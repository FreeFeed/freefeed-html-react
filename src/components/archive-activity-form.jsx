import PropTypes from 'prop-types';
import React from 'react';

export default class ArchiveActivityForm extends React.Component {
  static propTypes = {
    action: PropTypes.func.isRequired,
    formState: PropTypes.shape({
      inProgress: PropTypes.bool.isRequired,
      success: PropTypes.bool.isRequired,
      error: PropTypes.bool.isRequired,
      errorText: PropTypes.string.isRequired,
    }),
  };

  state = { checked: false };

  setChecked = (e) => this.setState({ checked: e.target.checked });

  action = (e) => {
    e.preventDefault();
    if (this.canSubmit()) {
      this.props.action();
    }
  };

  canSubmit() {
    return this.state.checked && !this.props.formState.inProgress;
  }

  render() {
    const { checked } = this.state;
    const { formState } = this.props;
    return (
      <form onSubmit={this.action}>
        <div className="checkbox">
          <label>
            <input type="checkbox" checked={checked} onClick={this.setChecked} />I allow to restore
            my comments and likes of other usersâ€™ posts
          </label>
        </div>
        <div className="form-group">
          <button type="submit" className="btn btn-default" disabled={!this.canSubmit()}>
            Yes, I allow
          </button>
        </div>
        {formState.error ? (
          <div className="alert alert-danger" role="alert">
            {formState.errorText}
          </div>
        ) : (
          false
        )}
      </form>
    );
  }
}
