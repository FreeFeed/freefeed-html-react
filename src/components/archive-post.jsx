import React from 'react';
import { connect } from 'react-redux';

function ArchivePostHandler({ router, inProgress, success, id }) {
  if (success) {
    router.replace(`/archive/${encodeURIComponent(id)}`);
    return null;
  }

  let postBody = null;

  if (inProgress) {
    postBody = (
      <p>
        Loading info for <strong>{router.location.query.url}</strong>...
      </p>
    );
  } else {
    postBody = (
      <div>
        <p>
          It seems that post <strong>{router.location.query.url}</strong> has not yet been restored
          from the archive.
        </p>
        <p>
          You can try to find{' '}
          <a
            target="_blank"
            style={{ textDecoration: 'underline' }}
            href={`https://web.archive.org/web/20150410000000/${encodeURIComponent(
              router.location.query.url,
            )}`}
          >
            copy of this post on archive.org
          </a>
          .
        </p>
      </div>
    );
  }

  return (
    <div className="box">
      <div className="box-header-timeline">Friendfeed Post</div>
      <div className="box-body">{postBody}</div>
      <div className="box-footer" />
    </div>
  );
}

function selectState(state) {
  const { archivePost } = state;
  return archivePost;
}

export default connect(selectState)(ArchivePostHandler);
