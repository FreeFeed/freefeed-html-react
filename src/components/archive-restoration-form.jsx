import PropTypes from 'prop-types';
import React from 'react';
import _ from 'lodash';
import { faExternalLinkAlt } from '@fortawesome/free-solid-svg-icons';
import { Icon } from './fontawesome-icons';

export default class ArchiveRestorationForm extends React.Component {
  static propTypes = {
    action: PropTypes.func.isRequired,
    formState: PropTypes.shape({
      inProgress: PropTypes.bool.isRequired,
      success: PropTypes.bool.isRequired,
      error: PropTypes.bool.isRequired,
      errorText: PropTypes.string.isRequired,
    }),
    sources: PropTypes.array.isRequired,
    oldUsername: PropTypes.string.isRequired,
  };

  state = {
    disable_comments: false,
    via_restore: [],
  };

  setDisableComments = (e) => this.setState({ disable_comments: e.target.checked });
  setViaSource = (e) => {
    const { checked, value } = e.target;
    if (checked) {
      this.setState({ via_restore: _.union(this.state.via_restore, [value]) });
    } else {
      this.setState({ via_restore: _.without(this.state.via_restore, value) });
    }
  };
  selectAllSources = (e) => {
    if (e.target.checked) {
      this.setState({ via_restore: this.props.sources.map((s) => s.url) });
    } else {
      this.setState({ via_restore: [] });
    }
  };

  canSubmit() {
    return this.state.via_restore.length > 0 && !this.props.formState.inProgress;
  }

  action = (e) => {
    e.preventDefault();
    if (this.canSubmit()) {
      this.props.action(this.state);
    }
  };

  render() {
    const { via_restore, disable_comments } = this.state;
    const { formState, sources, oldUsername } = this.props;
    const totalCount = sources.reduce((p, s) => p + s.count, 0);
    const selectedCount = sources
      .filter((s) => via_restore.includes(s.url))
      .reduce((p, s) => p + s.count, 0);

    const buttonText =
      totalCount === selectedCount
        ? 'Restore all my posts'
        : selectedCount === 0
        ? 'Restore my posts'
        : `Restore my ${selectedCount} ${selectedCount === 1 ? 'post' : 'posts'}`;

    return (
      <form onSubmit={this.action}>
        <p>
          You have {totalCount} posts in your <strong>friendfeed.com/{oldUsername}</strong> archive.
          Posts from which source would you like to restore?
        </p>
        <div className="checkbox">
          <label>
            <input
              type="checkbox"
              checked={via_restore.length === sources.length}
              onClick={this.selectAllSources}
            />
            Restore all {totalCount} posts (or select specific sources below)
          </label>
        </div>
        <SourceList sources={sources} selected={via_restore} onClick={this.setViaSource} />
        <div className="checkbox">
          <label>
            <input type="checkbox" checked={disable_comments} onClick={this.setDisableComments} />
            Disable new comments for your posts restored from archive (you can enable comments for
            individual posts after the restoring has been completed)
          </label>
        </div>
        <div className="form-group">
          <button type="submit" className="btn btn-default" disabled={!this.canSubmit()}>
            {buttonText}
          </button>
        </div>
        {formState.error ? (
          <div className="alert alert-danger" role="alert">
            {formState.errorText}
          </div>
        ) : (
          false
        )}
        <p>
          <em>Requests are handled manually, so be prepared to wait a bit.</em>
        </p>
      </form>
    );
  }
}

const FRF_URL = 'http://friendfeed.com';

class SourceList extends React.Component {
  static propTypes = {
    sources: PropTypes.array.isRequired,
    selected: PropTypes.array.isRequired,
    onClick: PropTypes.func.isRequired,
  };

  renderSource({ name, url }) {
    if (url === FRF_URL) {
      return <span>Posts on FriendFeed</span>;
    }
    if (url.indexOf(FRF_URL) === 0) {
      return <span>FriendFeed: {name}</span>;
    }
    return (
      <span>
        {name}
        <a href={url} title={url} target="_blank" className="source-link">
          <Icon icon={faExternalLinkAlt} />
        </a>
      </span>
    );
  }

  render() {
    const { selected, onClick } = this.props;
    const sources = this.props.sources.sort((a, b) => {
      if (a.url === FRF_URL) {
        return -1;
      }
      if (b.url === FRF_URL) {
        return 1;
      }
      return b.count - a.count;
    });
    return (
      <div className="archive-source-list">
        {sources.map((s) => (
          <div className="checkbox" key={s.url}>
            <label>
              <input
                type="checkbox"
                value={s.url}
                checked={selected.includes(s.url)}
                onClick={onClick}
              />
              {this.renderSource(s)} <span className="text-muted">({s.count})</span>
            </label>
          </div>
        ))}
      </div>
    );
  }
}
