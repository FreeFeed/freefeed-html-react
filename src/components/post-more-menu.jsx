import React from 'react'
import DropdownMenu from 'react-dd-menu';

export default class PostMoreMenu extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      isOpen: false
    }
  }

  toggle() {
    this.setState({ isOpen: !this.state.isOpen })
  }

  close() {
    this.setState({ isOpen: false })
  }

  render() {
    let menuOptions = {
      align: 'left',
      close: this.close.bind(this),
      isOpen: this.state.isOpen,
      toggle: <a onClick={this.toggle.bind(this)}>More&nbsp;&#x25be;</a>
    }

    return (
      <DropdownMenu {...menuOptions}>
        <li className="dd-menu-item"><a className="dd-menu-item-link" onClick={this.props.toggleEditingPost}>Edit</a></li>
        <li className="dd-menu-item dd-menu-item-danger"><a className="dd-menu-item-link" onClick={this.props.deletePost}>Delete</a></li>
      </DropdownMenu>
    )
  }
}
