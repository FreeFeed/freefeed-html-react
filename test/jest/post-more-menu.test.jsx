/* global jest, describe, it, expect */
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import * as reactRedux from 'react-redux';

import { PostMoreMenu } from '../../src/components/post-more-menu';

jest.mock('../../src/components/fontawesome-icons', () => ({
  Icon: ({ icon }) => `fontawesome icon ${icon.iconName}`, // mocking out icon to make snapshots smaller
}));

const VIEWER = {
  id: 'user-id',
  username: 'viewer',
  frontendPreferences: {
    timeDisplay: {},
  },
};

const POST = {
  id: 'post123',
  isEditable: false,
  canBeRemovedFrom: [],
  isModeratable: false,
  isDeletable: false,
  isModeratingComments: false,
  commentsDisabled: false,
  createdAt: '1629668000000',
  updatedAt: '1629668999999',
  isSaved: false,
  savePostStatus: {},
};

const defaultState = {
  user: VIEWER,
};

const renderPostMoreMenu = (props = {}) => {
  const { Provider } = reactRedux;
  const dummyReducer = (state) => state;
  const store = createStore(dummyReducer, defaultState);

  const defaultProps = {
    user: VIEWER,
    post: POST,
    toggleEditingPost: () => {},
    toggleModeratingComments: () => {},
    enableComments: () => {},
    disableComments: () => {},
    deletePost: () => {},
    perGroupDeleteEnabled: true,
    doAndClose: () => {},
    permalink: 'https://freefeed.net/post123',
    toggleSave: () => {},
  };

  return render(
    <Provider store={store}>
      <PostMoreMenu {...defaultProps} {...props} />
    </Provider>,
  );
};

describe('PostMoreMenu', () => {
  it('Renders a More menu for a logged-out visitor', () => {
    const { asFragment } = renderPostMoreMenu({ user: {} });

    expect(asFragment()).toMatchSnapshot();
  });

  it('Renders a More menu for a logged-in reader', () => {
    const { asFragment } = renderPostMoreMenu();

    expect(asFragment()).toMatchSnapshot();
  });

  it('Renders a More menu for a group moderator', () => {
    const { asFragment } = renderPostMoreMenu({
      post: {
        ...POST,
        isModeratable: true,
        canBeRemovedFrom: ['group1', 'group2'],
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('Renders a More menu for a post owner', () => {
    const { asFragment } = renderPostMoreMenu({
      post: {
        ...POST,
        isEditable: true,
        isDeletable: true,
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
