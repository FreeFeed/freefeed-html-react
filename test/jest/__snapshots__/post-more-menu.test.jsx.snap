// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostMoreMenu Renders a More menu for a group moderator 1`] = `
<DocumentFragment>
  <div
    class="list focusList"
    style="min-width: 18em;"
  >
    <div
      class="group"
    >
      <div
        class="item"
      >
        <a
          aria-disabled="false"
          class="link"
          role="button"
          tabindex="0"
        >
          <span
            class="iconic"
          >
            <span
              class="iconicIcon"
            >
              fontawesome icon comment-dots
            </span>
            <span
              class="iconicContent"
            >
              Moderate comments
            </span>
          </span>
        </a>
      </div>
      <div
        class="item"
      >
        <a
          aria-disabled="false"
          class="link"
          role="button"
          tabindex="0"
        >
          <span
            class="iconic"
          >
            <span
              class="iconicIcon"
            >
              fontawesome icon comment-dots
            </span>
            <span
              class="iconicContent"
            >
              Disable comments
            </span>
          </span>
        </a>
      </div>
    </div>
    <div
      class="group"
    >
      <div
        class="item"
      >
        <a
          aria-disabled="false"
          class="link danger"
          role="button"
          tabindex="0"
        >
          <span
            class="iconic"
          >
            <span
              class="iconicIcon"
            >
              fontawesome icon trash-alt
            </span>
            <span
              class="iconicContent"
            >
              Remove from @group1
            </span>
          </span>
        </a>
      </div>
      <div
        class="item"
      >
        <a
          aria-disabled="false"
          class="link danger"
          role="button"
          tabindex="0"
        >
          <span
            class="iconic"
          >
            <span
              class="iconicIcon"
            >
              fontawesome icon trash-alt
            </span>
            <span
              class="iconicContent"
            >
              Remove from @group2
            </span>
          </span>
        </a>
      </div>
    </div>
    <div
      class="group"
    >
      <div
        class="item content"
      >
        <span
          class="iconic"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon clock
          </span>
          <span
            class="iconicContent"
          >
            Created on 
            <time
              datetime="2021-08-22T21:33:20.000Z"
              title="Aug 22, 2021"
            >
              Aug 22, 2021 21:33
            </time>
          </span>
        </span>
      </div>
      <div
        class="item content"
      >
        <span
          class="iconic"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon clock
          </span>
          <span
            class="iconicContent"
          >
            Updated on 
            <time
              datetime="2021-08-22T21:49:59.999Z"
              title="Aug 22, 2021"
            >
              Aug 22, 2021 21:49
            </time>
          </span>
        </span>
      </div>
      <div
        class="item content"
      >
        <span
          class="iconic iconicCentered"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon link
          </span>
          <span
            class="iconicContent"
          >
            <a
              style="margin-right: 1ex;"
            >
              Link to post
            </a>
             
            <button
              aria-label="Copy link"
              class="btn btn-default btn-sm"
              type="button"
              value="https://freefeed.net/post123"
            >
              Copy
            </button>
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PostMoreMenu Renders a More menu for a logged-in reader 1`] = `
<DocumentFragment>
  <div
    class="list focusList"
    style="min-width: 18em;"
  >
    <div
      class="group"
    >
      <div
        class="item content"
      >
        <span
          class="iconic"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon clock
          </span>
          <span
            class="iconicContent"
          >
            Created on 
            <time
              datetime="2021-08-22T21:33:20.000Z"
              title="Aug 22, 2021"
            >
              Aug 22, 2021 21:33
            </time>
          </span>
        </span>
      </div>
      <div
        class="item content"
      >
        <span
          class="iconic"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon clock
          </span>
          <span
            class="iconicContent"
          >
            Updated on 
            <time
              datetime="2021-08-22T21:49:59.999Z"
              title="Aug 22, 2021"
            >
              Aug 22, 2021 21:49
            </time>
          </span>
        </span>
      </div>
      <div
        class="item content"
      >
        <span
          class="iconic iconicCentered"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon link
          </span>
          <span
            class="iconicContent"
          >
            <a
              style="margin-right: 1ex;"
            >
              Link to post
            </a>
             
            <button
              aria-label="Copy link"
              class="btn btn-default btn-sm"
              type="button"
              value="https://freefeed.net/post123"
            >
              Copy
            </button>
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PostMoreMenu Renders a More menu for a logged-out visitor 1`] = `
<DocumentFragment>
  <div
    class="list focusList"
    style="min-width: 18em;"
  >
    <div
      class="group"
    >
      <div
        class="item content"
      >
        <span
          class="iconic"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon clock
          </span>
          <span
            class="iconicContent"
          >
            Created on 
            <time
              datetime="2021-08-22T21:33:20.000Z"
              title="Aug 22, 2021"
            >
              Aug 22, 2021 21:33
            </time>
          </span>
        </span>
      </div>
      <div
        class="item content"
      >
        <span
          class="iconic"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon clock
          </span>
          <span
            class="iconicContent"
          >
            Updated on 
            <time
              datetime="2021-08-22T21:49:59.999Z"
              title="Aug 22, 2021"
            >
              Aug 22, 2021 21:49
            </time>
          </span>
        </span>
      </div>
      <div
        class="item content"
      >
        <span
          class="iconic iconicCentered"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon link
          </span>
          <span
            class="iconicContent"
          >
            <a
              style="margin-right: 1ex;"
            >
              Link to post
            </a>
             
            <button
              aria-label="Copy link"
              class="btn btn-default btn-sm"
              type="button"
              value="https://freefeed.net/post123"
            >
              Copy
            </button>
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PostMoreMenu Renders a More menu for a post owner 1`] = `
<DocumentFragment>
  <div
    class="list focusList"
    style="min-width: 18em;"
  >
    <div
      class="group"
    >
      <div
        class="item"
      >
        <a
          aria-disabled="false"
          class="link"
          role="button"
          tabindex="0"
        >
          <span
            class="iconic"
          >
            <span
              class="iconicIcon"
            >
              fontawesome icon edit
            </span>
            <span
              class="iconicContent"
            >
              Edit
            </span>
          </span>
        </a>
      </div>
      <div
        class="item"
      >
        <a
          aria-disabled="false"
          class="link"
          role="button"
          tabindex="0"
        >
          <span
            class="iconic"
          >
            <span
              class="iconicIcon"
            >
              fontawesome icon comment-dots
            </span>
            <span
              class="iconicContent"
            >
              Disable comments
            </span>
          </span>
        </a>
      </div>
    </div>
    <div
      class="group"
    >
      <div
        class="item"
      >
        <a
          aria-disabled="false"
          class="link danger"
          role="button"
          tabindex="0"
        >
          <span
            class="iconic"
          >
            <span
              class="iconicIcon"
            >
              fontawesome icon trash-alt
            </span>
            <span
              class="iconicContent"
            >
              Delete
            </span>
          </span>
        </a>
      </div>
    </div>
    <div
      class="group"
    >
      <div
        class="item content"
      >
        <span
          class="iconic"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon clock
          </span>
          <span
            class="iconicContent"
          >
            Created on 
            <time
              datetime="2021-08-22T21:33:20.000Z"
              title="Aug 22, 2021"
            >
              Aug 22, 2021 21:33
            </time>
          </span>
        </span>
      </div>
      <div
        class="item content"
      >
        <span
          class="iconic"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon clock
          </span>
          <span
            class="iconicContent"
          >
            Updated on 
            <time
              datetime="2021-08-22T21:49:59.999Z"
              title="Aug 22, 2021"
            >
              Aug 22, 2021 21:49
            </time>
          </span>
        </span>
      </div>
      <div
        class="item content"
      >
        <span
          class="iconic iconicCentered"
        >
          <span
            class="iconicIcon"
          >
            fontawesome icon link
          </span>
          <span
            class="iconicContent"
          >
            <a
              style="margin-right: 1ex;"
            >
              Link to post
            </a>
             
            <button
              aria-label="Copy link"
              class="btn btn-default btn-sm"
              type="button"
              value="https://freefeed.net/post123"
            >
              Copy
            </button>
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
